spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java

```diff
@@ -111,171 +49,17 @@ protected RewriteDataFilesAction self() {
   }
 
   @Override
-  protected Table table() {
-    return table;
-  }
-
-  /**
-   * Pass a PartitionSpec id to specify which PartitionSpec should be used in DataFile rewrite
-   *
-   * @param specId PartitionSpec id to rewrite
-   * @return this for method chaining
-   */
-  public RewriteDataFilesAction outputSpecId(int specId) {
-    Preconditions.checkArgument(table.specs().containsKey(specId), "Invalid spec id %d", specId);
-    this.spec = table.specs().get(specId);
-    return this;
-  }
-
-  /**
-   * Specify the target rewrite data file size in bytes
-   *
-   * @param targetSize size in bytes of rewrite data file
-   * @return this for method chaining
-   */
-  public RewriteDataFilesAction targetSizeInBytes(long targetSize) {
-    Preconditions.checkArgument(targetSize > 0L, "Invalid target rewrite data file size in bytes %d",
-        targetSize);
-    this.targetSizeInBytes = targetSize;
-    return this;
-  }
-
-  /**
-   * Specify the number of "bins" considered when trying to pack the next file split into a task.
-   * Increasing this usually makes tasks a bit more even by considering more ways to pack file regions into a single
-   * task with extra planning cost.
-   * <p>
-   * This configuration can reorder the incoming file regions, to preserve order for lower/upper bounds in file
-   * metadata, user can use a lookback of 1.
-   *
-   * @param lookback number of "bins" considered when trying to pack the next file split into a task.
-   * @return this for method chaining
-   */
-  public RewriteDataFilesAction splitLookback(int lookback) {
-    Preconditions.checkArgument(lookback > 0L, "Invalid split lookback %d", lookback);
-    this.splitLookback = lookback;
-    return this;
-  }
-
-  /**
-   * Specify the minimum file size to count to pack into one "bin". If the read file size is smaller than this specified
-   * threshold, Iceberg will use this value to do count.
-   * <p>
-   * this configuration controls the number of files to compact for each task, small value would lead to a
-   * high compaction, the default value is 4MB.
-   *
-   * @param openFileCost minimum file size to count to pack into one "bin".
-   * @return this for method chaining
-   */
-  public RewriteDataFilesAction splitOpenFileCost(long openFileCost) {
-    Preconditions.checkArgument(openFileCost > 0L, "Invalid split openFileCost %d", openFileCost);
-    this.splitOpenFileCost = openFileCost;
-    return this;
-  }
-
-  /**
-   * Pass a row Expression to filter DataFiles to be rewritten. Note that all files that may contain data matching the
-   * filter may be rewritten.
-   *
-   * @param expr Expression to filter out DataFiles
-   * @return this for method chaining
-   */
-  public RewriteDataFilesAction filter(Expression expr) {
-    this.filter = Expressions.and(filter, expr);
-    return this;
+  protected FileIO fileIO() {
+    return SparkUtil.serializableFileIO(table());
```

Comment:
should we make this lazy? 

private FileIO lazyFileIO;
protected FileIO fileIO { 
   if (lazyFileIO == null) { 
     lazyFileIO =...
   }
  return lazyFileIo
}
}

Reply:
we could make it lazy, I looked up the code which uses SparkUtil#serializableFileIO method, they are all directly call, so I think make it lazy may not make much sense.

Reply:
That would save lots of fileIO serialization if we make this lazy, I agree with @RussellSpitzer . The upper layer may call this fileIO method several times, while we only need to serialize once.

Reply:
This fileIO() method should be a setter for field fileIO. This method only needs to be called once when BaseRewriteDataFilesAction is initialized. Other places need to refer to fileIO and only need to call its getter method, 

If we rename fileIO() to setFileIO(), it might be better understood, but I don’t know if this conforms to iceberg's specification.

```
  // BaseRewriteDataFilesAction.java 
  protected FileIO fileIO() {
    return fileIO;
  }

  protected abstract FileIO setFileIO();
```

```
  // RewriteDataFilesAction.java 
  @Override
  protected FileIO setFileIO() {
    return SparkUtil.serializableFileIO(table());
  }
```

so I think the method will not be called multiple times. What do you think ？@openinx

Reply:
OK, sound like we've a local field to initialize the fileIO, this method will only call once now. I'm OK with it then.